graspologic/models/edge_swaps.py

```diff
@@ -0,0 +1,207 @@
+import warnings
+from asyncore import loop
+from typing import List, Optional, Union
+
+import networkx as nx
+import numba as nb
+import numpy as np
+from beartype import beartype
+from numba.core.errors import NumbaWarning
+from scipy.sparse import SparseEfficiencyWarning, csr_matrix
+
+from graspologic.preconditions import check_argument
+from graspologic.types import Tuple
+from graspologic.utils.utils import is_loopless, is_symmetric, is_unweighted
+
+
+# Code based on: https://github.com/joelnish/double-edge-swap-mcmc/blob/master/dbl_edge_mcmc.py
+class EdgeSwapper:
+    """
+    Degree Preserving Edge Swaps
+
+    This class allows for performing degree preserving edge swaps to
+    generate new networks with the same degree sequence as the input network.
+
+    Attributes
+    ----------
+    adjacency : np.ndarray OR csr_matrix, shape (n_verts, n_verts)
+        The initial adjacency matrix to perform edge swaps on. Must be unweighted and undirected.
+
+    edge_list : np.ndarray, shape (n_verts, 2)
+        The corresponding edgelist for the input network
+
+
+    References
+    ----------
+    .. [1] Fosdick, B. K., Larremore, D. B., Nishimura, J., & Ugander, J. (2018).
+           Configuring random graph models with fixed degree sequences.
+           Siam Review, 60(2), 315-355.
+
+    .. [2] Carstens, C. J., & Horadam, K. J. (2017).
+           Switching edges to randomize networks: what goes wrong and how to fix it.
+           Journal of Complex Networks, 5(3), 337-351.
+
+    .. [3] https://github.com/joelnish/double-edge-swap-mcmc/blob/master/dbl_edge_mcmc.py
+    """
+
+    @beartype
+    def __init__(self, adjacency: Union[np.ndarray, csr_matrix]):
```

Comment:
@kareef928 graspologic.types should have an AdjacencyMatrix type or something like that